File: python/tests/test_spec.py
"""The spec's account of what each language carries, held against what this side exports.

The last section of ``inst/spec/representation.md`` records every name on both sides, every name
shaped differently, and every name present in one language only. A public name it does not
record, or one it records on the wrong side, fails here, so the tables cannot fall behind the
exports. The R suite reads the same section against its own NAMESPACE.
"""

from __future__ import annotations

import re
from pathlib import Path

import timesift

SPEC = Path(__file__).resolve().parents[2] / "inst" / "spec" / "representation.md"
NAME = re.compile(r"`([A-Za-z_][A-Za-z0-9_.]*)(?:\(\))?`")


def section() -> list[str]:
    lines = SPEC.read_text(encoding="utf-8").splitlines()
    start = next(i for i, l in enumerate(lines) if l.startswith("## What each language carries"))
    return lines[start:]


def names(lines) -> set[str]:
    return {m for line in lines for m in NAME.findall(line)}


def table_names(lines, header: str, column: int = 0) -> set[str]:
    """The bare names in one column of the table whose header line is ``header``. A name written
    with an argument, such as ``elasticnet(s =)``, names an argument and is left out."""
    start = lines.index(header)
    cells = []
    for line in lines[start + 2:]:
        if not line.startswith("|"):
            break
        cells.append(line.lstrip("|").split("|")[column])
    return names(cells)


def test_every_public_name_is_recorded_in_the_spec():
    missing = set(timesift.__all__) - names(section())
    assert not missing, f"public but not in the spec: {sorted(missing)}"


def test_a_name_recorded_on_both_sides_is_public_here():
    shared = table_names(section(), "| concept | the name, on both sides |", column=1)
    shared.discard("n_inner")
    missing = shared - set(timesift.__all__)
    assert not missing, f"recorded on both sides but not public: {sorted(missing)}"


def test_a_name_recorded_on_one_side_is_on_that_side():
    lines = section()
    r_only = table_names(lines, "| in R only | why |")
    leaked = r_only & set(timesift.__all__)
    assert not leaked, f"recorded as R only but public here: {sorted(leaked)}"
    python_only = table_names(lines, "| in Python only | what it is |")
    missing = python_only - set(timesift.__all__)
    assert not missing, f"recorded as Python only but not public: {sorted(missing)}"

File: tests/testthat/test-spec.R
spec_section <- function() {
  path <- system.file("spec", "representation.md", package = "timesift")
  lines <- readLines(path, encoding = "UTF-8")
  lines[seq(grep("^## What each language carries", lines), length(lines))]
}

spec_names <- function(lines) {
  hits <- regmatches(lines, gregexpr("`[A-Za-z_][A-Za-z0-9_.]*(\\(\\))?`", lines))
  unique(sub("\\(\\)$", "", gsub("`", "", unlist(hits))))
}

# The first column of the table whose header is `header`, as the bare names it carries. A name
# written with an argument, such as `elasticnet(s =)`, names an argument rather than an export
# and is left out.
spec_table_names <- function(lines, header, column = 1L) {
  start <- grep(header, lines, fixed = TRUE)
  stopifnot(length(start) == 1L)
  rows <- character()
  for (i in seq(start + 2L, length(lines))) {
    if (!startsWith(lines[i], "|")) break
    cells <- strsplit(sub("^\\|", "", lines[i]), "|", fixed = TRUE)[[1L]]
    rows <- c(rows, cells[column])
  }
  spec_names(rows)
}

r_exports <- function() {
  ns <- readLines(system.file("NAMESPACE", package = "timesift"))
  sub("^export\\((.*)\\)$", "\\1", grep("^export\\(", ns, value = TRUE))
}

test_that("every export is recorded in the spec's account of what each language carries", {
  lines <- spec_section()
  missing <- setdiff(r_exports(), spec_names(lines))
  expect_true(!length(missing),
              info = paste("exported but not in the spec:", paste(missing, collapse = ", ")))
})

test_that("a name the spec records on both sides is exported here", {
  lines <- spec_section()
  shared <- spec_table_names(lines, "| concept | the name, on both sides |", column = 2L)
  shared <- shared[shared != "n_inner"]
  missing <- setdiff(shared, r_exports())
  expect_true(!length(missing),
              info = paste("recorded on both sides but not exported:",
                            paste(missing, collapse = ", ")))
})

test_that("a name the spec records as Python only is not an export here", {
  lines <- spec_section()
  python_only <- spec_table_names(lines, "| in Python only | what it is |")
  leaked <- intersect(python_only, r_exports())
  expect_true(!length(leaked),
              info = paste("recorded as Python only but exported:", paste(leaked, collapse = ", ")))
  r_only <- spec_table_names(lines, "| in R only | why |")
  r_only <- setdiff(r_only, "plot")
  missing <- setdiff(r_only, r_exports())
  expect_true(!length(missing),
              info = paste("recorded as R only but not exported:", paste(missing, collapse = ", ")))
})
